core/embed/trezorhal/stm32f4/usb/usbd_class_hid.c

```diff
@@ -337,18 +386,44 @@ static int usb_hid_class_setup(USBD_HandleTypeDef *dev, usb_hid_state_t *state,
   return USBD_OK;
 }
 
-static void usb_hid_class_data_in(USBD_HandleTypeDef *dev,
-                                  usb_hid_state_t *state, uint8_t ep_num) {
+static uint8_t usb_hid_class_data_in(USBD_HandleTypeDef *dev, uint8_t ep_num) {
+  usb_hid_state_t *state = (usb_hid_state_t *)dev->pUserData;
+
   if ((ep_num | USB_EP_DIR_IN) == state->ep_in) {
+    wait_random();
     state->ep_in_is_idle = 1;
   }
+
+  return USBD_OK;
 }
 
-static void usb_hid_class_data_out(USBD_HandleTypeDef *dev,
-                                   usb_hid_state_t *state, uint8_t ep_num) {
+static uint8_t usb_hid_class_data_out(USBD_HandleTypeDef *dev, uint8_t ep_num) {
+  usb_hid_state_t *state = (usb_hid_state_t *)dev->pUserData;
+
   if (ep_num == state->ep_out) {
+    wait_random();
     // Save the report length to indicate we have read something, but don't
     // schedule next reading until user reads this one
     state->last_read_len = USBD_LL_GetRxDataSize(dev, ep_num);
   }
+
+  return USBD_OK;
 }
+
+static const USBD_ClassTypeDef usb_hid_class = {
```

Comment:
The definition of `usb_hid_class` depends on functions defined in this file, so it must appear after them. What is provided here is just a forward declaration because `usb_hid_class` is needed by `usb_hid_add`. We could remove this forward declaration, but it would require restructuring the file content: placing the `usb_hid_class_xxx()` functions at the beginning and `usb_hid` at the end. However, I'm not sure if we want to do this. Perhaps I don't fully understand your point.

Reply:
Ah ok, makes sense. No need to restructure it.